File: graph_bfs.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

// Function to add an edge to our undirected graph
void addEdge(vector<vector<int>>& adj, int u, int v) {
    adj[u].push_back(v);
    adj[v].push_back(u);
}

// Core BFS Algorithm Function
void bfsGraph(const vector<vector<int>>& adj, int startNode, int V) {
    // 1. Ek boolean array banayein tracked vertices ke liye (Sab shuru mein false hain)
    vector<bool> visited(V, false);
    
    // 2. BFS ke liye ek standard queue banayein
    queue<int> q;

    // 3. Start node ko visited mark karein aur queue mein push karein
    visited[startNode] = true;
    q.push(startNode);

    cout << "BFS Traversal starting from Node " << startNode << ": ";

    while (!q.empty()) {
        // Queue ke samne wale element ko nikalie
        int curr = q.front();
        q.pop();
        
        cout << curr << " ";

        // Us current node ke saare neighbors ko check karein
        for (int neighbor : adj[curr]) {
            // Agar neighbor abhi tak visited nahi hai, toh use queue mein daalein
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
    cout << endl;
}

int main() {
    cout << "--- DSA Masterclass: Breadth-First Search (BFS) ---" << endl << endl;

    int V = 5; // 5 Nodes (0 se 4)
    vector<vector<int>> adj(V);

    // Graph exact pichle setup jaisa hi build karte hain
    addEdge(adj, 0, 1);
    addEdge(adj, 0, 4);
    addEdge(adj, 1, 2);
    addEdge(adj, 1, 3);
    addEdge(adj, 1, 4);
    addEdge(adj, 2, 3);
    addEdge(adj, 3, 4);

    // Node 0 ko starting point banakar BFS traverse karte hain
    bfsGraph(adj, 0, V);

    cout << "---------------------------------------------------" << endl;
    return 0;
}

File: graph_bfs.py
from collections import deque

# Graph क्लास जो Adjacency List का इस्तेमाल करती है
class Graph:
    def __init__(self):
        # ग्राफ को स्टोर करने के लिए डिक्शनरी
        self.graph = {}

    # ग्राफ में दो नोड्स के बीच कनेक्शन (Edge) जोड़ने के लिए
    def add_edge(self, u, v):
        if u not in self.graph:
            self.graph[u] = []
        if v not in self.graph:
            self.graph[v] = []
            
        self.graph[u].append(v)
        self.graph[v].append(u)  # Undirected Graph के लिए दोनों तरफ जोड़ेंगे

    # BFS Traversal फंक्शन
    def bfs(self, start_node):
        visited = set()         # ट्रैक रखने के लिए कि कौन सा नोड विजिट हो चुका है
        queue = deque()         # नोड्स को प्रोसेस करने के लिए कतार

        # शुरुआती नोड से शुरुआत करें
        visited.add(start_node)
        queue.append(start_node)

        print(f"BFS Traversal starting from node {start_node}: ", end="")

        while queue:
            # क्यू के सामने से एलिमेंट निकालें
            current = queue.popleft()
            print(f"{current} -> ", end="")

            # करंट नोड के सभी पड़ोसियों को चेक करें
            for neighbor in self.graph[current]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        
        print("End")


# --- Main Driver Code ---
if __name__ == "__main__":
    g = Graph()
    
    print("--- Creating the Graph ---")
    # एक सिंपल ग्राफ का स्ट्रक्चर बना रहे हैं
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    g.add_edge(1, 4)
    g.add_edge(2, 4)

    # BFS रन करें
    g.bfs(0)